File: cliffy-protocols/src/consensus.rs
//! Geometric consensus protocol implementations

use crate::{geometric_mean, serde_ga3, GeometricCRDT, OperationType};
use cliffy_core::GA3;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::{broadcast, RwLock};
use uuid::Uuid;

/// A message in the consensus protocol
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConsensusMessage {
    pub sender_id: Uuid,
    #[serde(with = "serde_ga3")]
    pub proposal: GA3,
    pub round: u64,
    pub message_type: MessageType,
}

/// Types of consensus messages
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum MessageType {
    Propose(#[serde(with = "serde_ga3")] GA3),
    Vote(bool, #[serde(with = "serde_ga3")] GA3),
    Commit(#[serde(with = "serde_ga3")] GA3),
    Sync(GeometricCRDT),
}

/// A geometric consensus protocol implementation
pub struct GeometricConsensus {
    node_id: Uuid,
    current_round: u64,
    proposals: Arc<RwLock<HashMap<u64, Vec<GA3>>>>,
    votes: Arc<RwLock<HashMap<u64, HashMap<Uuid, bool>>>>,
    committed_states: Arc<RwLock<HashMap<u64, GA3>>>,
    message_sender: broadcast::Sender<ConsensusMessage>,
    message_receiver: broadcast::Receiver<ConsensusMessage>,
    crdt_state: Arc<RwLock<GeometricCRDT>>,
}

impl GeometricConsensus {
    /// Create a new consensus protocol instance
    pub fn new(node_id: Uuid, initial_state: GA3) -> Self {
        let (sender, receiver) = broadcast::channel(1000);
        let crdt = GeometricCRDT::new(node_id, initial_state);

        Self {
            node_id,
            current_round: 0,
            proposals: Arc::new(RwLock::new(HashMap::new())),
            votes: Arc::new(RwLock::new(HashMap::new())),
            committed_states: Arc::new(RwLock::new(HashMap::new())),
            message_sender: sender,
            message_receiver: receiver,
            crdt_state: Arc::new(RwLock::new(crdt)),
        }
    }

    /// Propose a value for consensus
    pub async fn propose(&mut self, value: GA3) -> Result<(), Box<dyn std::error::Error>> {
        let round = self.current_round;
        self.current_round += 1;

        let message = ConsensusMessage {
            sender_id: self.node_id,
            proposal: value.clone(),
            round,
            message_type: MessageType::Propose(value),
        };

        self.message_sender.send(message)?;
        Ok(())
    }

    /// Compute consensus from a set of proposals using geometric algebra
    pub async fn geometric_consensus(
        &self,
        proposals: &[GA3],
        threshold: f64,
    ) -> Result<GA3, Box<dyn std::error::Error>> {
        if proposals.is_empty() {
            return Ok(GA3::zero());
        }

        // Compute geometric mean of all proposals
        let consensus_value = geometric_mean(proposals);

        // Check if consensus meets threshold (based on geometric distance)
        let max_distance = proposals
            .iter()
            .map(|proposal| {
                let diff = &consensus_value - proposal;
                diff.magnitude()
            })
            .fold(0.0_f64, |acc, dist| acc.max(dist));

        if max_distance <= threshold {
            Ok(consensus_value)
        } else {
            // No consensus reached, return weighted geometric mean
            self.weighted_geometric_consensus(proposals).await
        }
    }

    /// Compute weighted geometric consensus based on magnitudes
    async fn weighted_geometric_consensus(
        &self,
        proposals: &[GA3],
    ) -> Result<GA3, Box<dyn std::error::Error>> {
        // Weight proposals by their geometric magnitude
        let weights: Vec<f64> = proposals.iter().map(|p| p.magnitude()).collect();

        let total_weight: f64 = weights.iter().sum();

        if total_weight == 0.0 {
            return Ok(GA3::zero());
        }

        // Simple weighted average for now
        let mut result = GA3::zero();
        for (proposal, weight) in proposals.iter().zip(weights.iter()) {
            let scaled: Vec<f64> = proposal
                .as_slice()
                .iter()
                .map(|&c| c * weight / total_weight)
                .collect();
            let scaled_mv = GA3::from_slice(&scaled);
            result = &result + &scaled_mv;
        }

        Ok(result)
    }

    /// Run a full consensus round
    pub async fn run_consensus_round(
        &mut self,
        proposal: GA3,
        participants: &[Uuid],
    ) -> Result<Option<GA3>, Box<dyn std::error::Error>> {
        let round = self.current_round;

        // Phase 1: Propose
        self.propose(proposal.clone()).await?;

        // Collect proposals from all participants
        let mut received_proposals = vec![proposal];
        let mut proposal_count = 1;

        while proposal_count < participants.len() {
            if let Ok(message) = self.message_receiver.recv().await {
                if message.round == round {
                    if let MessageType::Propose(prop) = message.message_type {
                        received_proposals.push(prop);
                        proposal_count += 1;
                    }
                }
            }
        }

        // Phase 2: Vote
        let consensus_candidate = self.geometric_consensus(&received_proposals, 0.1).await?;
        let vote_message = ConsensusMessage {
            sender_id: self.node_id,
            proposal: consensus_candidate.clone(),
            round,
            message_type: MessageType::Vote(true, consensus_candidate.clone()),
        };

        self.message_sender.send(vote_message)?;

        // Collect votes
        let mut votes = HashMap::new();
        votes.insert(self.node_id, true);

        while votes.len() < participants.len() {
            if let Ok(message) = self.message_receiver.recv().await {
                if message.round == round {
                    if let MessageType::Vote(vote, _) = message.message_type {
                        votes.insert(message.sender_id, vote);
                    }
                }
            }
        }

        // Phase 3: Commit if majority votes yes
        let yes_votes = votes.values().filter(|&&v| v).count();
        if yes_votes > participants.len() / 2 {
            let commit_message = ConsensusMessage {
                sender_id: self.node_id,
                proposal: consensus_candidate.clone(),
                round,
                message_type: MessageType::Commit(consensus_candidate.clone()),
            };

            self.message_sender.send(commit_message)?;

            // Update CRDT state
            let mut crdt_guard = self.crdt_state.write().await;
            let op = crdt_guard.create_operation(consensus_candidate.clone(), OperationType::Addition);
            crdt_guard.apply_operation(op);

            Ok(Some(consensus_candidate))
        } else {
            Ok(None)
        }
    }

    /// Sync CRDT state with another node
    pub async fn sync_crdt_state(
        &self,
        _other_node: Uuid,
    ) -> Result<(), Box<dyn std::error::Error>> {
        let crdt_guard = self.crdt_state.read().await;
        let sync_message = ConsensusMessage {
            sender_id: self.node_id,
            proposal: crdt_guard.state.clone(),
            round: self.current_round,
            message_type: MessageType::Sync(crdt_guard.clone()),
        };

        self.message_sender.send(sync_message)?;
        Ok(())
    }

    /// Handle an incoming sync message
    pub async fn handle_sync_message(&self, crdt_state: GeometricCRDT) {
        let mut local_crdt = self.crdt_state.write().await;
        *local_crdt = local_crdt.merge(&crdt_state);
    }

    /// Get the current consensus state
    pub async fn get_current_state(&self) -> GA3 {
        let crdt_guard = self.crdt_state.read().await;
        crdt_guard.state.clone()
    }
}

/// Compute the lattice join (least upper bound) of two multivectors
pub fn lattice_join(a: &GA3, b: &GA3) -> GA3 {
    let a_coeffs = a.as_slice();
    let b_coeffs = b.as_slice();

    let result_coeffs: Vec<f64> = a_coeffs
        .iter()
        .zip(b_coeffs.iter())
        .map(|(&ac, &bc)| ac.max(bc))
        .collect();

    GA3::from_slice(&result_coeffs)
}

/// Compute the lattice meet (greatest lower bound) of two multivectors
pub fn lattice_meet(a: &GA3, b: &GA3) -> GA3 {
    let a_coeffs = a.as_slice();
    let b_coeffs = b.as_slice();

    let result_coeffs: Vec<f64> = a_coeffs
        .iter()
        .zip(b_coeffs.iter())
        .map(|(&ac, &bc)| ac.min(bc))
        .collect();

    GA3::from_slice(&result_coeffs)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_geometric_consensus_simple() {
        let proposals = vec![
            GA3::scalar(1.0),
            GA3::scalar(2.0),
            GA3::scalar(4.0),
        ];

        let node_id = Uuid::new_v4();
        let consensus = GeometricConsensus::new(node_id, GA3::zero());

        let result = consensus
            .geometric_consensus(&proposals, 5.0) // Use larger threshold for test
            .await
            .unwrap();

        // Result should be some weighted average
        assert!(result.get(0) > 0.0);
    }

    #[test]
    fn test_lattice_operations() {
        let a = GA3::from_slice(&[1.0, 2.0, 3.0, 0.0, 0.0, 0.0, 0.0, 0.0]);
        let b = GA3::from_slice(&[2.0, 1.0, 4.0, 0.0, 0.0, 0.0, 0.0, 0.0]);

        let join = lattice_join(&a, &b);
        let meet = lattice_meet(&a, &b);

        assert!((join.get(0) - 2.0).abs() < 1e-10);
        assert!((join.get(1) - 2.0).abs() < 1e-10);
        assert!((join.get(2) - 4.0).abs() < 1e-10);

        assert!((meet.get(0) - 1.0).abs() < 1e-10);
        assert!((meet.get(1) - 1.0).abs() < 1e-10);
        assert!((meet.get(2) - 3.0).abs() < 1e-10);
    }
}

File: cliffy-tsukoshi/src/protocols/consensus.ts
/**
 * Geometric consensus protocol implementations.
 *
 * Uses geometric algebra for conflict-free consensus in distributed systems.
 * The geometric mean provides a mathematically elegant way to combine
 * proposals while maintaining convergence guarantees.
 *
 * @example
 * ```typescript
 * import { GeometricConsensus, latticeJoin, latticeMeet } from 'cliffy-tsukoshi/protocols';
 * import { scalar, fromCoefficients } from 'cliffy-tsukoshi';
 *
 * const nodeId = crypto.randomUUID();
 * const consensus = new GeometricConsensus(nodeId, scalar(0));
 *
 * // Propose a value
 * consensus.propose(scalar(5));
 *
 * // Compute consensus from multiple proposals
 * const result = consensus.geometricConsensus([
 *   scalar(1), scalar(2), scalar(4)
 * ], 0.1);
 * ```
 */

import type { GA3 } from '../ga3.js';
import {
  zero,
  magnitude,
  sub,
  add,
  fromCoefficients,
} from '../ga3.js';
import { GeometricCRDT, OperationType, geometricMean } from './crdt.js';

/**
 * Types of consensus messages.
 */
export type MessageType =
  | { type: 'Propose'; value: GA3 }
  | { type: 'Vote'; accept: boolean; value: GA3 }
  | { type: 'Commit'; value: GA3 }
  | { type: 'Sync'; crdt: GeometricCRDT };

/**
 * A message in the consensus protocol.
 */
export interface ConsensusMessage {
  senderId: string;
  proposal: GA3;
  round: number;
  messageType: MessageType;
}

/**
 * Event handler for consensus messages.
 */
export type ConsensusMessageHandler = (message: ConsensusMessage) => void;

/**
 * A geometric consensus protocol implementation.
 *
 * Uses geometric mean for proposal combining and supports both
 * synchronous and asynchronous consensus patterns.
 */
export class GeometricConsensus {
  private nodeId: string;
  private currentRound: number;
  private proposals: Map<number, GA3[]>;
  private votes: Map<number, Map<string, boolean>>;
  private committedStates: Map<number, GA3>;
  private crdtState: GeometricCRDT;
  private messageHandlers: ConsensusMessageHandler[];

  constructor(nodeId: string, initialState: GA3) {
    this.nodeId = nodeId;
    this.currentRound = 0;
    this.proposals = new Map();
    this.votes = new Map();
    this.committedStates = new Map();
    this.crdtState = new GeometricCRDT(nodeId, initialState);
    this.messageHandlers = [];
  }

  /**
   * Subscribe to outgoing consensus messages.
   */
  onMessage(handler: ConsensusMessageHandler): () => void {
    this.messageHandlers.push(handler);
    return () => {
      const index = this.messageHandlers.indexOf(handler);
      if (index >= 0) {
        this.messageHandlers.splice(index, 1);
      }
    };
  }

  /**
   * Broadcast a message to all handlers.
   */
  private broadcast(message: ConsensusMessage): void {
    for (const handler of this.messageHandlers) {
      handler(message);
    }
  }

  /**
   * Propose a value for consensus.
   */
  propose(value: GA3): number {
    const round = this.currentRound++;

    const message: ConsensusMessage = {
      senderId: this.nodeId,
      proposal: value,
      round,
      messageType: { type: 'Propose', value },
    };

    this.broadcast(message);
    return round;
  }

  /**
   * Compute consensus from a set of proposals using geometric algebra.
   */
  geometricConsensus(proposals: GA3[], threshold: number): GA3 {
    if (proposals.length === 0) {
      return zero();
    }

    // Compute geometric mean of all proposals
    const consensusValue = geometricMean(proposals);

    // Check if consensus meets threshold (based on geometric distance)
    let maxDistance = 0;
    for (const proposal of proposals) {
      const diff = sub(consensusValue, proposal);
      const dist = magnitude(diff);
      maxDistance = Math.max(maxDistance, dist);
    }

    if (maxDistance <= threshold) {
      return consensusValue;
    } else {
      // No consensus reached, return weighted geometric mean
      return this.weightedGeometricConsensus(proposals);
    }
  }

  /**
   * Compute weighted geometric consensus based on magnitudes.
   */
  private weightedGeometricConsensus(proposals: GA3[]): GA3 {
    // Weight proposals by their geometric magnitude
    const weights = proposals.map(p => magnitude(p));
    const totalWeight = weights.reduce((a, b) => a + b, 0);

    if (totalWeight === 0) {
      return zero();
    }

    // Weighted average
    let result = zero();
    for (let i = 0; i < proposals.length; i++) {
      const coeffs = proposals[i] as number[];
      const scaledCoeffs = coeffs.map(c => c * weights[i] / totalWeight);
      result = add(result, fromCoefficients(scaledCoeffs));
    }

    return result;
  }

  /**
   * Handle an incoming proposal.
   */
  receiveProposal(_senderId: string, value: GA3, round: number): void {
    if (!this.proposals.has(round)) {
      this.proposals.set(round, []);
    }
    this.proposals.get(round)!.push(value);
  }

  /**
   * Cast a vote on a proposal.
   */
  vote(round: number, accept: boolean, value: GA3): void {
    if (!this.votes.has(round)) {
      this.votes.set(round, new Map());
    }
    this.votes.get(round)!.set(this.nodeId, accept);

    const message: ConsensusMessage = {
      senderId: this.nodeId,
      proposal: value,
      round,
      messageType: { type: 'Vote', accept, value },
    };

    this.broadcast(message);
  }

  /**
   * Handle an incoming vote.
   */
  receiveVote(senderId: string, round: number, accept: boolean): void {
    if (!this.votes.has(round)) {
      this.votes.set(round, new Map());
    }
    this.votes.get(round)!.set(senderId, accept);
  }

  /**
   * Try to commit a round if we have enough votes.
   */
  tryCommit(round: number, participantCount: number): GA3 | null {
    const roundVotes = this.votes.get(round);
    if (!roundVotes) {
      return null;
    }

    const yesVotes = Array.from(roundVotes.values()).filter(v => v).length;

    if (yesVotes > participantCount / 2) {
      const proposals = this.proposals.get(round) ?? [];
      const consensusValue = this.geometricConsensus(proposals, 0.1);

      // Commit
      this.committedStates.set(round, consensusValue);

      // Update CRDT state
      const op = this.crdtState.createOperation(consensusValue, OperationType.Addition);
      this.crdtState.applyOperation(op);

      const message: ConsensusMessage = {
        senderId: this.nodeId,
        proposal: consensusValue,
        round,
        messageType: { type: 'Commit', value: consensusValue },
      };

      this.broadcast(message);
      return consensusValue;
    }

    return null;
  }

  /**
   * Handle an incoming commit.
   */
  receiveCommit(value: GA3, round: number): void {
    if (!this.committedStates.has(round)) {
      this.committedStates.set(round, value);

      // Update CRDT state
      const op = this.crdtState.createOperation(value, OperationType.Addition);
      this.crdtState.applyOperation(op);
    }
  }

  /**
   * Sync CRDT state with another node.
   */
  syncCrdtState(): void {
    const message: ConsensusMessage = {
      senderId: this.nodeId,
      proposal: this.crdtState.state,
      round: this.currentRound,
      messageType: { type: 'Sync', crdt: this.crdtState },
    };

    this.broadcast(message);
  }

  /**
   * Handle an incoming sync message.
   */
  handleSyncMessage(crdtState: GeometricCRDT): void {
    this.crdtState = this.crdtState.merge(crdtState);
  }

  /**
   * Get the current consensus state.
   */
  getCurrentState(): GA3 {
    return this.crdtState.state;
  }

  /**
   * Get the underlying CRDT.
   */
  getCrdt(): GeometricCRDT {
    return this.crdtState;
  }

  /**
   * Get the current round number.
   */
  getCurrentRound(): number {
    return this.currentRound;
  }

  /**
   * Get proposals for a round.
   */
  getProposals(round: number): GA3[] {
    return this.proposals.get(round) ?? [];
  }

  /**
   * Get votes for a round.
   */
  getVotes(round: number): Map<string, boolean> {
    return this.votes.get(round) ?? new Map();
  }

  /**
   * Check if a round has been committed.
   */
  isCommitted(round: number): boolean {
    return this.committedStates.has(round);
  }

  /**
   * Get committed value for a round.
   */
  getCommittedValue(round: number): GA3 | null {
    return this.committedStates.get(round) ?? null;
  }
}

/**
 * Compute the component-wise lattice join (least upper bound).
 */
export function latticeJoin(a: GA3, b: GA3): GA3 {
  const aCoeffs = a as number[];
  const bCoeffs = b as number[];
  const result = aCoeffs.map((c, i) => Math.max(c, bCoeffs[i]));
  return fromCoefficients(result);
}

/**
 * Compute the component-wise lattice meet (greatest lower bound).
 */
export function latticeMeet(a: GA3, b: GA3): GA3 {
  const aCoeffs = a as number[];
  const bCoeffs = b as number[];
  const result = aCoeffs.map((c, i) => Math.min(c, bCoeffs[i]));
  return fromCoefficients(result);
}
